src/Tools/Common/TraceFilter.cs

```diff
@@ -0,0 +1,269 @@
+            providerNameExcludePattern = $"{sanitizedPattern.ToLowerInvariant()}{(string.IsNullOrEmpty(providerNameExcludePattern) ? "" : $"|{providerNameExcludePattern}")}";
+            return this;
+        }
+
+        public PredicateBuilder IncludeProviderPattern(string pattern)
+        {
+            string sanitizedPattern = pattern.Replace("*", ".*");
+            SimpleLogger.Log.WriteLine($"Adding include pattern: '{sanitizedPattern}'");
+            providerNameIncludePattern = $"{sanitizedPattern.ToLowerInvariant()}{(string.IsNullOrEmpty(providerNameIncludePattern) ? "" : $"|{providerNameIncludePattern}")}";
+            return this;
+        }
+
+        public PredicateBuilder AddProviderFilter(string providerName, string eventNamePattern)
+        {
+            string key = providerName.ToLowerInvariant();
+            if (!providerPredicates.ContainsKey(key))
+                providerPredicates[key] = new ProviderPredicate(key);
+
+            SimpleLogger.Log.WriteLine($"Adding event name filter: Provider={providerName}, pattern={eventNamePattern}");
+
+            providerPredicates[key].IncludeEventName(eventNamePattern);
+            return this;
+        }
+
+        public PredicateBuilder AddProviderFilter(string providerName, int eventId)
+        {
+            string key = providerName.ToLowerInvariant();
+            if (!providerPredicates.ContainsKey(key))
+                providerPredicates[key] = new ProviderPredicate(key);
+
+            SimpleLogger.Log.WriteLine($"Adding event id filter: Provider={providerName}, id={eventId}");
+
+            providerPredicates[key].IncludeEventId(eventId);
+            return this;
+        }
+
+        public PredicateBuilder AddProviderFilter(string providerName, long keyword)
+        {
+            string key = providerName.ToLowerInvariant();
+            if (!providerPredicates.ContainsKey(key))
+                providerPredicates[key] = new ProviderPredicate(key);
+
+            SimpleLogger.Log.WriteLine($"Adding event keyword filter: Provider={providerName}, keyword={keyword:X}");
+
+            providerPredicates[key].IncludeKeyword(keyword);
+            return this;
+        }
+
+        public Func<TraceEvent, bool> Build()
+        {
+            Func<TraceEvent, bool> predicate = (_) => true;
+
+            if (string.IsNullOrEmpty(providerNameIncludePattern) && string.IsNullOrEmpty(providerNameExcludePattern) && providerPredicates.Count == 0)
+                return predicate;
+
+            // order of checking:
+            // 1. exclude regex
+            // 2. include regex
+            // 3. subfilters
+            // effectively: exclude & include & (subfilters OR'd together)
+            predicate = (TraceEvent data) =>
+            {
+                bool ret = true;
+                if (!string.IsNullOrEmpty(providerNameExcludePattern))
+                    ret &= !Regex.IsMatch(data.ProviderName.ToLowerInvariant(), providerNameExcludePattern);
```

Comment:
For filters, rather than checking whether some `TraceEvent` meets a list of all conditions and then returning whether it succeeded/failed to meet all of them, wouldn't it be cheaper to return false upon not meeting a condition immediately, and return true at the end of all these checks?